File: 2_Regression/ARX_Regression/apm_id.py
import numpy as np
from apm import *

def apm_id(data,ni,nu,ny):
    fast = False  # switch "False" for solution analysis
    obj_scale = 1
    n = np.size(data,0)
    p = np.size(data,1)
    no = p - ni - 1
    if no<=0:
        print('Data file needs at least 1 column for output data')
        return
    m = max(ny,nu)

    # first column is time
    time = data[:,0].copy()
    dt = time[1] - time[0]

    # inputs are the next column(s)
    u_ss = data[0,1:1+ni].copy()
    u = data[:,1:1+ni].copy()

    # outputs are the final column(s)
    y_ss = data[0,1+ni:1+ni+no].copy()
    y = data[:,1+ni:1+ni+no].copy()

    for i in range(n):
        for j in range(ni):
            u[i,j] = u[i,j] - u_ss[j]
        for j in range(no):
            y[i,j] = y[i,j] - y_ss[j]

    fid = open('myModel.apm','w')
    fid.write('Objects\n')
    fid.write('  sum_a[1:no] = sum(%i)\n'%ny)
    fid.write('  sum_b[1:ni][1::no] = sum(%i)\n'%nu)
    fid.write('End Objects\n')
    fid.write('  \n')
    fid.write('Connections\n')
    fid.write('  a[1:ny][1::no] = sum_a[1::no].x[1:ny]\n')
    fid.write('  b[1:nu][1::ni][1:::no] = sum_b[1::ni][1:::no].x[1:nu]\n')
    fid.write('  sum_a[1:no] = sum_a[1:no].y\n')
    fid.write('  sum_b[1:ni][1::no] = sum_b[1:ni][1::no].y\n')
    fid.write('End Connections\n')
    fid.write('  \n')
    fid.write('Constants\n')
    fid.write('  n = %s\n' %str(n))
    fid.write('  ni = %s\n'%str(ni))
    fid.write('  no = %s\n'%str(no))
    fid.write('  ny = %s\n'%str(ny))
    fid.write('  nu = %s\n'%str(nu))
    fid.write('  m = %s\n'%str(m))
    fid.write('  \n')
    fid.write('Parameters\n')
    fid.write('  a[1:ny][1::no] = 0 !>= -1 <= 1\n')
    fid.write('  b[1:nu][1::ni][1:::no] = 0\n')
    fid.write('  c[1:no] = 0\n')
    fid.write('  u[1:n][1::ni]\n')
    fid.write('  y[1:m][1::no]\n')
    fid.write('  z[1:n][1::no]\n')
    fid.write('  \n')
    fid.write('Variables\n')
    fid.write('  y[m+1:n][1::no] = 0\n')
    fid.write('  sum_a[1:no] = 0 !<= 1\n')
    fid.write('  sum_b[1:ni][1::no] = 0\n')
    fid.write('  K[1:ni][1::no] = 0 !>=0 <= 10 \n') 
    fid.write('  \n')
    fid.write('Equations\n')
    fid.write('  y[m+1:n][1::no] = a[1][1::no]*y[m:n-1][1::no]')
    for j in range(1,ni+1):
        fid.write(' + b[1][%i][1::no]*u[m:n-1][%i]'%(j,j,))
        for i in range(2,nu+1): 
            fid.write(' + b[%i][%i][1::no]*u[m-%i:n-%i][%i]'%(i,j,i-1,i,j,))
    for i in range(2,ny+1): 
        fid.write(' + a[%i][1::no]*y[m-%i:n-%i][1::no]'%(i,i-1,i,))
    fid.write('\n')
    fid.write('  K[1:ni][1::no] * (1 - sum_a[1::no]) = sum_b[1:ni][1::no]\n')
    fid.write('  minimize %e * (y[m+1:n][1::no] - z[m+1:n][1::no])^2\n'%obj_scale)
    fid.close()
    
    fid = open('myData.csv','w')
    for j in range(1,ni+1): 
        for i in range(1,n+1): 
            fid.write('u[%i][%i], %e\n'%(i,j,u[i-1,j-1],))
    for k in range(1,no+1):
        for i in range(1,n+1):
            fid.write('z[%i][%i], %e\n'%(i,k,y[i-1,k-1],))
    for k in range(1,no+1): 
        for i in range(1,n+1): 
            fid.write('y[%i][%i], %e\n'%(i,k,y[i-1,k-1],))
    fid.close()
    
    s = 'http://byu.apmonitor.com'
    #s = 'http://127.0.0.1'
    a = 'apm_id'
    apm(s,a,'clear all')
    apm_load(s,a,'myModel.apm')
    csv_load(s,a,'myData.csv')
    
    apm_option(s,a,'nlc.solver',3)
    apm_option(s,a,'nlc.imode',2)
    apm_option(s,a,'nlc.max_iter',100)
    
    for i in range(1,no+1): 
        name = 'c[' + str(i) + ']'
        apm_info(s,a,'FV',name)
        apm_option(s,a,name+'.status',0)
    for k in range(1,no+1): 
        for i in range(1,ny+1): 
            name = 'a[' + str(i) + '][' + str(k) + ']'
            apm_info(s,a,'FV',name)
            apm_option(s,a,name+'.status',1)
    for k in range(1,no+1): 
        for j in range(1,nu+1): 
            for i in range(1,nu+1): 
                name = 'b[' + str(i) + '][' + str(j) + '][' + str(k) + ']'
                apm_info(s,a,'FV',name)
                apm_option(s,a,name+'.status',1)
    output = apm(s,a,'solve')

    # retrieve and visualize solution
    sol = apm_sol(s,a)
    ypred = np.empty((n,no))
    for j in range(no):
        for i in range(n):
            yn = 'y['+str(i+1)+']['+str(j+1)+']'
            ypred[i,j] = sol[yn]

    if (not fast):
        # open web-viewer
        apm_web(s,a)

        import matplotlib.pyplot as plt
        plt.figure()
        plt.subplot(2,1,1)
        for i in range(ni):
            plt.plot(time,u[:,i]+u_ss[i])
        plt.subplot(2,1,2)
        for i in range(no):
            plt.plot(time,y[:,i]+y_ss[i],'r-')
            plt.plot(time,ypred[:,i]+y_ss[i],'b--')
        plt.show()

    success = apm_tag(s,a,'nlc.appstatus')
    if success==1:
        obj = apm_tag(s,a,'nlc.objfcnval')
        if (not fast):
            print(output)
        print('Successful solution with objective: ' + str(obj))
    else:
        print(output)
        print('Unsuccessful, do not write sysa.apm')
        return

    alpha = np.empty((ny,no))
    beta = np.empty((nu,ni,no))
    gamma = np.empty((no))
    for j in range(1,no+1):
        for i in range(1,ny+1):
            name = 'a['+str(i)+']['+str(j)+']'
            alpha[i-1,j-1] = apm_tag(s,a,name+'.newval');
    for k in range(1,no+1):
        for j in range(1,ni+1):
            for i in range(1,nu+1):
                name = 'b['+str(i)+']['+str(j)+']['+str(k)+']'
                beta[i-1,j-1,k-1] = apm_tag(s,a,name+'.newval')
    for i in range(1,no+1):
        name = 'c['+str(i)+']'
        gamma[i-1] = apm_tag(s,a,name+'.newval')

    name = 'sysa'
    fid = open('sysa.apm','w')
    fid.write('Objects\n')
    fid.write(' '+name+' = arx\n')    
    fid.write('End Objects\n')
    fid.write('\n')    
    fid.write('Connections\n')
    if ni==1:
        fid.write('  u = '+name+'.u\n')
    else:
        fid.write('  u[1:%i] = '%(ni,)+name+'.u[1:%i]\n'%(ni,))
    if no ==1:
        fid.write('  y = '+name+'.y\n')
    else:
        fid.write('  y[1:%i] = '%(no,)+name+'.y[1:%i]\n'%(no,))
    fid.write('End Connections\n')
    fid.write('\n')
    fid.write('Model \n')
    fid.write('  Parameters \n')
    if ni==1:
        fid.write('    u = 0\n')
    else:
        fid.write('    u[1:%i] = 0\n'%(ni,))
    fid.write('  End Parameters \n')
    fid.write('\n')
    fid.write('  Variables \n')
    if no==1:
        fid.write('    y = 0\n')
    else:
        fid.write('    y[1:%i] = 0\n'%(no,))
    fid.write('  End Variables \n')
    fid.write('\n')
    fid.write('  Equations \n')
    fid.write('    ! add any additional equations here \n')
    fid.write('  End Equations \n')
    fid.write('End Model \n')
    fid.write('\n')
    fid.write('File '+name+'.txt\n')
    fid.write('  %i      ! m=number of inputs\n'%(ni,))
    fid.write('  %i      ! p=number of outputs\n'%(no,))
    fid.write('  %i      ! nu=number of input terms\n'%(nu,))
    fid.write('  %i      ! ny=number of output terms\n'%(ny,))
    fid.write('End File\n')
    fid.write('\n')
    fid.write('! Alpha matrix (ny x p)\n')
    fid.write('File '+name+'.alpha.txt \n')
    for i in range(1,ny+1):
        for j in range(1,no+1):
            fid.write('  ')
            if j<=no-1:
                fid.write('%e , '%(alpha[i-1,j-1],))
            else:
                fid.write('%e '%(alpha[i-1,j-1],)) # no comma
        fid.write('\n')
    fid.write('End File \n')
    fid.write('\n')
    fid.write('! Beta matrix (p x (nu x m))\n')
    fid.write('File '+name+'.beta.txt \n')
    for i in range(1,no+1):
        for j in range(1,nu+1):
            fid.write('  ')
            for k in range(1,ni+1):
                if k<=ni-1:
                    fid.write('%e , '%(beta[j-1,k-1,i-1],))
                else:
                    fid.write('%e '%(beta[j-1,k-1,i-1],))
            fid.write('\n')
    fid.write('End File \n')
    fid.write('\n')
    fid.write('! Gamma vector (p x 1)\n')
    fid.write('File '+name+'.gamma.txt \n')
    for i in range(1,no+1):
        fid.write('%e \n'%(gamma[i-1],))
    fid.write('End File \n')
    fid.close()

    msg = 'Created ARX (Auto-Regressive eXogenous inputs) Model: sysa.apm'
    print(msg)

    return ypred+y_ss

File: 5_Moving_Horizon_Estimation/1st_order_linear/MATLAB/apm/apm_id.m
% Function apm_id for identification of a linear model
% Inputs
%   data = data set with time, inputs, outputs
%   ni = number of inputs (others are outputs)
%   nu = number of input terms (numerator)
%   ny = number of output terms (denominator)
%
% Outputs
%   sysd = discrete transfer function estimated from data
function sysd = apm_id(data,ni,nu,ny)

%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
% Configuration
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
% include plots and other analysis if it does not need to be fast
fast = false;
obj_scale = 1;

% size of data
[n,p] = size(data);
no = p - ni - 1;
if (no<=0)
    error('Data file needs additional columns for outputs')
end
m = max(ny,nu);

% time
time = data(:,1);
% time step
dt = time(2) - time(1);
% input
u_ss = data(1,2:1+ni);
u = data(:,2:1+ni);
% output
y_ss = data(1,2+ni:1+ni+no);
y = data(:,2+ni:1+ni+no);
for i = 1:size(y,1),
    u(i,:) = u(i,:) - u_ss;
    y(i,:) = y(i,:) - y_ss;
end

%% write model
fid = fopen('myModel.apm','w');
fprintf(fid,'Objects\n');
fprintf(fid,'  sum_a[1:no] = sum(%i)\n',ny);
fprintf(fid,'  sum_b[1:ni][1::no] = sum(%i)\n',nu);
fprintf(fid,'End Objects\n');
fprintf(fid,'  \n');
fprintf(fid,'Connections\n');
fprintf(fid,'  a[1:ny][1::no] = sum_a[1::no].x[1:ny]\n');
fprintf(fid,'  b[1:nu][1::ni][1:::no] = sum_b[1::ni][1:::no].x[1:nu]\n');
fprintf(fid,'  sum_a[1:no] = sum_a[1:no].y\n');
fprintf(fid,'  sum_b[1:ni][1::no] = sum_b[1:ni][1::no].y\n');
fprintf(fid,'End Connections\n');
fprintf(fid,'  \n');
fprintf(fid,'Constants\n');
fprintf(fid,'  n = %s\n',int2str(n));
fprintf(fid,'  ni = %s\n',int2str(ni));
fprintf(fid,'  no = %s\n',int2str(no));
fprintf(fid,'  ny = %s\n',int2str(ny));
fprintf(fid,'  nu = %s\n',int2str(nu));
fprintf(fid,'  m = %s\n',int2str(m));
fprintf(fid,'  \n');
fprintf(fid,'Parameters\n');
fprintf(fid,'  a[1:ny][1::no] = 0 >= -1 <= 1\n');
fprintf(fid,'  b[1:nu][1::ni][1:::no] = 0\n');
fprintf(fid,'  c[1:no] = 0\n');
fprintf(fid,'  u[1:n][1::ni]\n');
fprintf(fid,'  y[1:m][1::no]\n');
fprintf(fid,'  z[1:n][1::no]\n');
fprintf(fid,'  \n');
fprintf(fid,'Variables\n');
fprintf(fid,'  y[m+1:n][1::no] = 0\n');
fprintf(fid,'  sum_a[1:no] = 0 , <= 1\n');
fprintf(fid,'  sum_b[1:ni][1::no] = 0\n');
fprintf(fid,'  K[1:ni][1::no] = 0 >= 0 <= 10\n');
fprintf(fid,'  \n');
fprintf(fid,'Equations\n');
fprintf(fid,'  y[m+1:n][1::no] = a[1][1::no]*y[m:n-1][1::no]');
for j = 1:ni
    fprintf(fid,' + b[1][%i][1::no]*u[m:n-1][%i]',j,j);
    for i = 2:nu
        fprintf(fid, ' + b[%i][%i][1::no]*u[m-%i:n-%i][%i]',i,j,i-1,i,j);
    end
end
for i = 2:ny
    fprintf(fid, ' + a[%i][1::no]*y[m-%i:n-%i][1::no]',i,i-1,i);
end
fprintf(fid,'\n');
%        K(i,j) = sum(beta(:,j,i)) / (1-sum(alpha(:,i)));
fprintf(fid,'  K[1:ni][1::no] * (1 - sum_a[1::no]) = sum_b[1:ni][1::no]\n');
fprintf(fid,'  minimize %12.8f * (y[m+1:n][1::no] - z[m+1:n][1::no])^2\n',obj_scale);
fclose(fid);

%% write data
fid = fopen('myData.csv','w');
for j = 1:ni
    for i = 1:n
        fprintf(fid,'u[%i][%i], %12.8f\n', i,j,u(i,j));
    end
end
for k = 1:no
    for i = 1:n
        fprintf(fid,'z[%i][%i], %12.8f\n', i,k,y(i,k));
    end
end
for k = 1:no
    for i = 1:n
        fprintf(fid,'y[%i][%i], %12.8f\n', i,k,y(i,k));
    end
end
fclose(fid);

%% set up
addpath('apm');
s = 'http://byu.apmonitor.com';
a = 'apm_id';
apm(s,a,'clear all');
apm_load(s,a,'myModel.apm');
csv_load(s,a,'myData.csv');

apm_option(s,a,'nlc.solver',3);
apm_option(s,a,'nlc.imode',2);
apm_option(s,a,'nlc.max_iter',100);

% estimate nominal steady state conditions
%  c = y0 - sum_(i=1)^ny (ai * y0) - sum_(i=1)^nu (bi * u0)
for i = 1:no
    name = ['c[' int2str(i) ']'];
    apm_info(s,a,'FV',name);
    apm_option(s,a,[name '.status'],0);
end

% estimate alphas
for k = 1:no
    for i = 1:ny
        name = ['a[' int2str(i) '][' int2str(k) ']'];
        apm_info(s,a,'FV',name);
        apm_option(s,a,[name '.status'],1);
    end
end
% estimate betas
for k = 1:no
    for j = 1:ni
        for i = 1:nu
            name = ['b[' int2str(i) '][' int2str(j) '][' int2str(k) ']'];
            apm_info(s,a,'FV',name);
            apm_option(s,a,[name '.status'],1);
        end
    end
end
%% solve
output = apm(s,a,'solve');
disp(output);


%% retrieve and visualize solution
sol = apm_sol(s,a);
for j = 1:no
    for i = 1:n
        eval(['ypred(' int2str(i) ',' int2str(j) ') = sol.x.y' int2str(i) int2str(j) ';']);
    end
end
for j = 1:no
    for i = 1:ny
        name = ['a[' int2str(i) '][' int2str(j) ']'];
        alpha(i,j) = apm_tag(s,a,[name '.newval']);
    end
end
for k = 1:no
    for j = 1:ni
        for i = 1:nu
            name = ['b[' int2str(i) '][' int2str(j) '][' int2str(k) ']'];
            beta(i,j,k) = apm_tag(s,a,[name '.newval']);
        end
    end
end
for i = 1:no
    name = ['c[' int2str(i) ']'];  
    gamma(i) = apm_tag(s,a,[name '.newval']);
end

for i = 1:no
    for j = 1:ni
        K(i,j) = sum(beta(:,j,i)) / (1-sum(alpha(:,i)));
    end
end

if (~fast),
    disp('alpha:')
    disp(alpha)
    disp('beta:')
    disp(beta)
    disp('gamma:')
    disp(gamma)
    disp('K')
    disp(K)
    
    % plot results
    figure(1)
    
    subplot(3,1,1)
    plot(time,u(:,1),'r-','LineWidth',2);
    
    subplot(3,1,2)
    plot(time,u(:,2),'k:','LineWidth',2);
    
    subplot(3,1,3)
    plot(time,y,'b--','LineWidth',2);
    hold on
    plot(time,ypred);
    
    apm_web(s,a); 
end

% generate ARX model
apm_arx(no,ni,ny,nu,alpha,beta,gamma,'sysa');

% generate different model forms
% MATLAB doesn't like zero roots
for i = 1:ny,
    for j = 1:no,
        if (abs(alpha(i,j))<1e-5),
            alpha(i,j) = 1e-5;
        end
    end
end
% create LTI model, discrete transfer function
for i = 1:no,
    for j = 1:ni,
        sysd(i,j) = tf([0 beta(:,j,i)'],[1 -alpha(:,i)'],dt,'Variable','z^-1');
    end
end

% generate APMonitor model files (sysc.apm and sysd.apm)
% discrete state space models
apm_lti(sysd,'sysd');
if (~fast)
    % generate continuous state space form
    sysc = d2c(sysd);
    apm_lti(sysd,'sysc');
    figure(2)
    step(sysd,sysc)
end

return